docs/source/model_library.rst

```diff
@@ -0,0 +1,550 @@
+.. _model_library:
+
+Model Library
+=============
+
+`~stpipe.library.AbstractModelLibrary` is a container designed to allow efficient processing of
+collections of `~stpipe.datamodel.AbstractDataModel` instances created from an association.
+
+`~stpipe.library.AbstractModelLibrary` is an ordered collection (like a `list`) but provides:
+
+- access to association metadata: `~stpipe.library.AbstractModelLibrary.asn`
+- grouping API: `~stpipe.library.AbstractModelLibrary.group_indices` and `~stpipe.library.AbstractModelLibrary.group_names`
+- compatibility with `~stpipe.step.Step` and `~stpipe.pipeline.Pipeline` runs
+- a consistent indexing API that is the same for "in memory" and "on disk" libraries
+
+
+.. _library_association:
+
+Loading an association
+----------------------
+
+Most commonly an instance will be created from an association file:
+
+.. code-block:: pycon
+
+   >>> library = ModelLibrary("my_asn.json")
+
+.. NOTE::
+   For an association that contains a ``group_id`` for each member,
+   creating a library will not read any models.
+
+.. _library_borrowing_and_shelving:
+
+Borrowing and shelving models
+-----------------------------
+
+Interacting with an `~stpipe.library.AbstractModelLibrary` involves "borrowing" and "shelving"
+models, both of which must occur during a ``with`` statement (while the library
+is "open"):
+
+.. code-block:: pycon
+
+   >>> with library:
+   ...    model = library.borrow(0)
+   ...    # do stuff with the model...
+   ...    library.shelve(model)
+
+Iteration is also supported (but don't forget to return your models!).
+
+.. code-block:: pycon
+
+   >>> with library:
+   ...    for model in library:
+   ...        # do stuff with the model...
+   ...        library.shelve(model)
+
+
+.. _library_on_disk:
+
+On Disk Mode
+------------
+
+For large associations (like those larger than memory) it is important
+that the library avoid reading all models at once. The borrow/shelve API
+above maps closely to the loading/saving of input (or tempoary) files
+containing the models.
+
+.. code-block:: pycon
+
+   >>> library = ModelLibrary("my_big_asn.json", on_disk=True)
+   >>> with library:
+   ...     model = library.borrow(0)  # the input file for model 0 is loaded
+   ...     library.shelve(model)  # a temporary file for model 0 is written
+
+.. NOTE::
+   In the above example, a temporary file was created for model 0. At no
+   point will the library overwrite the input file.
+
```

Comment:
it might be good to note somewhere here that `model` is still in memory, e.g.,
```
with library:
    model = library.borrow(0)
    library.shelve(model, 0)
#do something with model
```
works.  This was unintuitive to me at first, the fact that the shelve command is NOT equivalent to `del model` in a case like this.

Reply:
Good point! How does the warning added here look to you?: b484e29